### ISSUE TITLE
getting scRepertoire across different celltypes 

### ISSUE THREAD

**Issue by jjs847** [ncborcherding/scRepertoire#340]
Hi @ncborcherding,

first, just want to mention I really appreciate your packages and documentation!!!

I want to look at repertoires across different B cell subsets (like naive vs post-switch memory for example).  
I have been using createHTOContigList() and grouping by both HTO_maxID and celltype so I have a separate list for each hashtag (which is time point data) and celltype combination. Then, depending on the celltype of interest, when I combineBCR I set samples to the vector for the 4 hashes/time points for the celltype, like samples = c("hto 1.naive", "hto 2.naive", "hto 3.naive", "hto 4.naive"). I then make plots based on these samples, however, I noticed that no matter which cell type all outputs look the same. Additionally, for some samples I get this error:
> combined <- combineBCR(contig.list)
Error in `bind_rows()`:
! Can't combine `..1$IGH` <character> and `..5$IGH` <logical>.
Backtrace:
 1. scRepertoire::combineBCR(contig.list)

I would have liked to use the AddVariable function but its not barcode specific, so to use this would you recommend creating a vector that is in the same order as the contiglist? However, I wasn't sure since I wanted to look at the repertoire in a specific cell subset if I have to run combineBCR for that subset separately.

To summarize, what I would like to do is look at the repertoire in 1 celltype across the 4 hashes, for example using clonal compare to visualize clonal expansion in naive B cells across the 4 time points. 

Thanks in advance for your time. 

Best, 

Jasmine

**Comment by ncborcherding** [comment]
Hey Jasmine, 

Thanks for reaching out and happy to help. 

Could you give me a little more information - 

- What is the output of ```sessionInfo()```?
- What does your ```contig.list``` look like? like ```length()``` and ```dim(contig.list[[1]])```, ```head(contig.list[[1]])```
- What is the plot outputs?

Once we figure the above out, I think it will be easier to get the ```addVariable()``` to work.

Thanks,
Nick

**Comment by jjs847** [comment]
Hi Nick, 

here is the requested information, however, I'm not sure I am approaching this correctly so I also summarize my issue below 
```
> contig.list_pt12_hto_celltype <- createHTOContigList(pt12, obj, group.by = c("HTO_maxID", "celltype"), multi.run = "patient_experiment")
> length(contig.list_pt12_hto_celltype)
[1] 40
> length(contig.list_pt12_hto_celltype[[1]])
[1] 31
> dim(contig.list_pt12_hto_celltype[[1]])
[1]  0 31
> head(contig.list_pt12_hto_celltype[[1]])
 [1] barcode               is_cell              
 [3] contig_id             high_confidence      
 [5] length                chain                
 [7] v_gene                d_gene               
 [9] j_gene                c_gene               
[11] full_length           productive           
[13] fwr1                  fwr1_nt              
[15] cdr1                  cdr1_nt              
[17] fwr2                  fwr2_nt              
[19] cdr2                  cdr2_nt              
[21] fwr3                  fwr3_nt              
[23] cdr3                  cdr3_nt              
[25] fwr4                  fwr4_nt              
[27] reads                 umis                 
[29] raw_clonotype_id      raw_consensus_id     
[31] exact_subclonotype_id
<0 rows> (or 0-length row.names)
> length(contig.list_pt12_hto_celltype[[3]])
[1] 31
> dim(contig.list_pt12_hto_celltype[[3]])
[1]  2 31
> head(contig.list_pt12_hto_celltype[[3]])
                     barcode is_cell                   contig_id
21632 TGGACGCAGTTAGCGG-1_4_2    true TGGACGCAGTTAGCGG-1_contig_1
21633 TGGACGCAGTTAGCGG-1_4_2    true TGGACGCAGTTAGCGG-1_contig_2
      high_confidence length chain   v_gene  d_gene j_gene c_gene
21632            true    653   IGL IGLV1-47          IGLJ2  IGLC2
21633            true    592   IGH IGHV3-23 IGHD2-2  IGHJ6   IGHM
      full_length productive                      fwr1
21632        true       true    QSVLTQPPSASGTPGQRVTISC
21633        true       true EVQLLESGGGLVQPGGSLRLSCAAS
                                                                          fwr1_nt
21632          CAGTCTGTGCTGACTCAGCCACCCTCAGCGTCTGGGACCCCCGGGCAGAGGGTCACCATCTCTTGT
21633 GAGGTGCAGCTGTTGGAGTCTGGGGGAGGCTTGGTACAGCCTGGGGGGTCCCTGAGACTCTCCTGTGCAGCCTCT
               cdr1                                 cdr1_nt
21632 SGSSSNIGSNYVY TCTGGAAGCAGCTCCAACATCGGAAGTAATTATGTATAC
21633       GFTFSSY                   GGATTCACCTTTAGCAGCTAT
                     fwr2
21632     WYQQLPGTAPKLLIY
21633 AMSWVRQAPGKGLEWVSAI
                                                        fwr2_nt
21632             TGGTACCAGCAGCTCCCAGGAACGGCCCCCAAACTCCTCATCTAT
21633 GCCATGAGCTGGGTCCGCCAGGCTCCAGGGAAGGGGCTGGAGTGGGTCTCAGCTATT
         cdr2               cdr2_nt
21632 RNNQRPS AGGAATAATCAGCGGCCCTCA
21633  SGSGGS    AGTGGTAGTGGTGGTAGC
                                        fwr3
21632        GVPDRFSGSKSGTSASLAISGLRSEDEADYY
21633 TYYADSVKGRFTISRDNSKNTLYLQMNSLRAEDTAVYY
                                                                                                                 fwr3_nt
21632                      GGGGTCCCTGACCGATTCTCTGGCTCCAAGTCTGGCACCTCAGCCTCCCTGGCCATCAGTGGGCTCCGGTCCGAGGATGAGGCTGATTATTAC
21633 ACATACTACGCAGACTCCGTGAAGGGCCGGTTCACCATCTCCAGAGACAATTCCAAGAACACGCTGTATCTGCAAATGAACAGCCTGAGAGCCGAGGACACGGCCGTATATTAC
                         cdr3
21632          CAAWDDSLSGHVVF
21633 CAKEDCSSTSCYGVYYYYGMDVW
                                                                    cdr3_nt
21632                            TGTGCAGCATGGGATGACAGCCTGAGTGGTCATGTGGTATTC
21633 TGTGCGAAAGAGGATTGTAGTAGTACCAGCTGCTATGGGGTCTACTACTACTACGGTATGGACGTCTGG
            fwr4                         fwr4_nt reads umis
21632  GGGTKLTVL    GGCGGAGGGACCAAGCTGACCGTCCTAG  1001   18
21633 GQGTTVTVSS GGCCAAGGGACCACGGTCACCGTCTCCTCAG   278    9
      raw_clonotype_id           raw_consensus_id
21632   clonotype10786 clonotype10786_consensus_2
21633   clonotype10786 clonotype10786_consensus_1
      exact_subclonotype_id
21632                     1
21633                     1
> sessionInfo()
R version 4.2.2 (2022-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Red Hat Enterprise Linux 8.6 (Ootpa)

Matrix products: default
BLAS:   /gpfs/share/apps/R/4.2.2/lib64/R/lib/libRblas.so
LAPACK: /gpfs/share/apps/R/4.2.2/lib64/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] grid      stats     graphics  grDevices utils     datasets 
[7] methods   base     

other attached packages:
 [1] patchwork_1.2.0       scRepertoire_2.0.0   
 [3] gridExtra_2.3         circlize_0.4.16      
 [5] tibble_3.2.1          tidyr_1.3.1          
 [7] ComplexHeatmap_2.14.0 ggrepel_0.9.5        
 [9] reshape2_1.4.4        dplyr_1.1.4          
[11] Azimuth_0.5.0         shinyBS_0.61.1       
[13] ggplot2_3.5.0         Seurat_5.0.3         
[15] SeuratObject_5.0.1    sp_2.1-3             

loaded via a namespace (and not attached):
  [1] rappdirs_0.3.3                   
  [2] SparseM_1.81                     
  [3] rtracklayer_1.58.0               
  [4] scattermore_1.2                  
  [5] iNEXT_3.0.0                      
  [6] R.methodsS3_1.8.2                
  [7] evmix_2.12                       
  [8] JASPAR2020_0.99.10               
  [9] bit64_4.0.5                      
 [10] knitr_1.45                       
 [11] irlba_2.3.5.1                    
 [12] DelayedArray_0.24.0              
 [13] R.utils_2.12.3                   
 [14] data.table_1.15.2                
 [15] doParallel_1.0.17                
 [16] KEGGREST_1.38.0                  
 [17] TFBSTools_1.36.0                 
 [18] RCurl_1.98-1.14                  
 [19] AnnotationFilter_1.22.0          
 [20] generics_0.1.3                   
 [21] BiocGenerics_0.44.0              
 [22] GenomicFeatures_1.50.4           
 [23] cowplot_1.1.3                    
 [24] RSQLite_2.3.5                    
 [25] VGAM_1.1-10                      
 [26] RANN_2.6.1                       
 [27] future_1.33.1                    
 [28] bit_4.0.5                        
 [29] tzdb_0.4.0                       
 [30] spatstat.data_3.0-4              
 [31] xml2_1.3.6                       
 [32] httpuv_1.6.14                    
 [33] SummarizedExperiment_1.28.0      
 [34] DirichletMultinomial_1.40.0      
 [35] viridis_0.6.5                    
 [36] gargle_1.5.2                     
 [37] xfun_0.42                        
 [38] hms_1.1.3                        
 [39] evaluate_0.23                    
 [40] promises_1.2.1                   
 [41] fansi_1.0.6                      
 [42] restfulr_0.0.15                  
 [43] progress_1.2.3                   
 [44] caTools_1.18.2                   
 [45] dbplyr_2.5.0                     
 [46] igraph_2.0.3                     
 [47] DBI_1.2.2                        
 [48] htmlwidgets_1.6.4                
 [49] spatstat.geom_3.2-9              
 [50] stringdist_0.9.12                
 [51] googledrive_2.1.1                
 [52] stats4_4.2.2                     
 [53] purrr_1.0.2                      
 [54] ellipsis_0.3.2                   
 [55] RSpectra_0.16-1                  
 [56] annotate_1.76.0                  
 [57] biomaRt_2.54.1                   
 [58] deldir_2.0-4                     
 [59] MatrixGenerics_1.10.0            
 [60] ggalluvial_0.12.5                
 [61] SingleCellExperiment_1.20.1      
 [62] vctrs_0.6.5                      
 [63] Biobase_2.58.0                   
 [64] quantreg_5.97                    
 [65] SeuratDisk_0.0.0.9021            
 [66] ensembldb_2.22.0                 
 [67] ROCR_1.0-11                      
 [68] abind_1.4-5                      
 [69] cachem_1.0.8                     
 [70] withr_3.0.0                      
 [71] ggforce_0.4.2                    
 [72] BSgenome.Hsapiens.UCSC.hg38_1.4.5
 [73] BSgenome_1.66.3                  
 [74] progressr_0.14.0                 
 [75] presto_1.0.0                     
 [76] sctransform_0.4.1                
 [77] GenomicAlignments_1.34.1         
 [78] prettyunits_1.2.0                
 [79] goftest_1.2-3                    
 [80] cluster_2.1.6                    
 [81] gsl_2.1-8                        
 [82] dotCall64_1.1-1                  
 [83] lazyeval_0.2.2                   
 [84] seqLogo_1.64.0                   
 [85] crayon_1.5.2                     
 [86] hdf5r_1.3.10                     
 [87] spatstat.explore_3.2-7           
 [88] labeling_0.4.3                   
 [89] pkgconfig_2.0.3                  
 [90] tweenr_2.0.3                     
 [91] GenomeInfoDb_1.39.5              
 [92] nlme_3.1-164                     
 [93] ProtGenerics_1.30.0              
 [94] rlang_1.1.3                      
 [95] globals_0.16.3                   
 [96] lifecycle_1.0.4                  
 [97] miniUI_0.1.1.1                   
 [98] MatrixModels_0.5-3               
 [99] filelock_1.0.3                   
[100] fastDummies_1.7.3                
[101] BiocFileCache_2.6.1              
[102] SeuratData_0.2.2.9001            
[103] cellranger_1.1.0                 
[104] polyclip_1.10-6                  
[105] RcppHNSW_0.6.0                   
[106] matrixStats_1.2.0                
[107] lmtest_0.9-40                    
[108] Matrix_1.6-5                     
[109] Rhdf5lib_1.20.0                  
[110] zoo_1.8-12                       
[111] GlobalOptions_0.1.2              
[112] ggridges_0.5.6                   
[113] googlesheets4_1.1.1              
[114] png_0.1-8                        
[115] viridisLite_0.4.2                
[116] rjson_0.2.21                     
[117] shinydashboard_0.7.2             
[118] bitops_1.0-7                     
[119] R.oo_1.26.0                      
[120] rhdf5filters_1.10.1              
[121] KernSmooth_2.23-22               
[122] spam_2.10-0                      
[123] Biostrings_2.66.0                
[124] blob_1.2.4                       
[125] shape_1.4.6.1                    
[126] stringr_1.5.1                    
[127] parallelly_1.37.1                
[128] spatstat.random_3.2-3            
[129] readr_2.1.5                      
[130] S4Vectors_0.36.2                 
[131] CNEr_1.34.0                      
[132] scales_1.3.0                     
[133] memoise_2.0.1                    
[134] magrittr_2.0.3                   
[135] plyr_1.8.9                       
[136] ica_1.0-3                        
[137] zlibbioc_1.44.0                  
[138] compiler_4.2.2                   
[139] BiocIO_1.8.0                     
[140] RColorBrewer_1.1-3               
[141] clue_0.3-65                      
[142] fitdistrplus_1.1-11              
[143] Rsamtools_2.14.0                 
[144] cli_3.6.2                        
[145] XVector_0.38.0                   
[146] listenv_0.9.1                    
[147] pbapply_1.7-2                    
[148] mgcv_1.9-1                       
[149] MASS_7.3-60.0.1                  
[150] tidyselect_1.2.1                 
[151] stringi_1.8.3                    
[152] yaml_2.3.8                       
[153] fastmatch_1.1-4                  
[154] EnsDb.Hsapiens.v86_2.99.0        
[155] tools_4.2.2                      
[156] future.apply_1.11.1              
[157] parallel_4.2.2                   
[158] rstudioapi_0.15.0                
[159] TFMPvalue_0.0.9                  
[160] foreach_1.5.2                    
[161] cubature_2.1.0                   
[162] farver_2.1.1                     
[163] Rtsne_0.17                       
[164] ggraph_2.2.1                     
[165] digest_0.6.35                    
[166] shiny_1.8.0                      
[167] pracma_2.4.4                     
[168] Rcpp_1.0.12                      
[169] GenomicRanges_1.50.2             
[170] later_1.3.2                      
[171] RcppAnnoy_0.0.22                 
[172] ggdendro_0.2.0                   
[173] httr_1.4.7                       
[174] AnnotationDbi_1.60.2             
[175] colorspace_2.1-0                 
[176] XML_3.99-0.16.1                  
[177] fs_1.6.3                         
[178] tensor_1.5                       
[179] reticulate_1.35.0                
[180] IRanges_2.32.0                   
[181] splines_4.2.2                    
[182] uwot_0.1.16                      
[183] RcppRoll_0.3.0                   
[184] spatstat.utils_3.0-4             
[185] graphlayouts_1.1.1               
[186] plotly_4.10.4                    
[187] xtable_1.8-4                     
[188] truncdist_1.0-2                  
[189] jsonlite_1.8.8                   
[190] poweRlaw_0.80.0                  
[191] tidygraph_1.3.1                  
[192] R6_2.5.1                         
[193] pillar_1.9.0                     
[194] htmltools_0.5.7                  
[195] mime_0.12                        
[196] glue_1.7.0                       
[197] fastmap_1.1.1                    
[198] DT_0.32                          
[199] BiocParallel_1.32.6              
[200] codetools_0.2-19                 
[201] Signac_1.12.0                    
[202] utf8_1.2.4                       
[203] lattice_0.22-6                   
[204] spatstat.sparse_3.0-3            
[205] evd_2.3-6.1                      
[206] curl_5.2.0                       
[207] leiden_0.4.3.1                   
[208] gtools_3.9.5                     
[209] shinyjs_2.1.0                    
[210] GO.db_3.16.0                     
[211] survival_3.5-8                   
[212] rmarkdown_2.26                   
[213] munsell_0.5.0                    
[214] GetoptLong_1.0.5                 
[215] rhdf5_2.42.1                     
[216] GenomeInfoDbData_1.2.9           
[217] iterators_1.0.14                 
[218] gtable_0.3.4 
```
It makes sense that for the first list it doesn't work since there is no data. 
After I have the contig list and split by cell type and hashtag, I combineBCR for a cell type of interest as such:
```
combined_bcr_pt12 <- combineBCR(contig.list_pt12_hto_celltype, samples = c("hto 1.DN2", "hto 2.DN2", "hto 3.DN2", "hto 4.DN2"))
```
since there are 40 slots with the hashtag-celltype combination, I call the specific celltype I am interested in in the samples arguement. 

then, I subset the combinedBCR output and visualize the celltype of interest:
```
keep_names <- c("hto 1.DN2", "hto 2.DN2", "hto 3.DN2", "hto 4.DN2")
combined_bcr_pt1 <- combined_bcr_pt1[names(combined_bcr_pt1) %in% keep_names]
  
p1 = clonalQuant(combined_bcr_pt1, cloneCall = "nt") + NoLegend()
p2 = clonalHomeostasis(scr, cloneCall = "nt")
p1_2 = p1 + p2
  
p3 = clonalCompare(combined_bcr_pt1, top.clones = 50, samples = c("hto 1.DN2", "hto 2.DN2", "hto 3.DN2", "hto 4.DN2"), cloneCall="nt", graph = "alluvial") + NoLegend()
  
```

I do this for multiple cell types. What I am confused about here is that when I changed the string of samples in combineBCR e.g., "combinedBCR(contig.list_pt12_hto_celltype, samples = c("hto 1.naive", "hto 2.naive", "hto 3.naive", "hto 4.naive")), the outputs are all identical no matter the celltype. 

I think I am approaching this incorrectly. **The TDLR of what i'd like to do is visualize the repertoire across hashtags for a specific celltype.** My understanding of the addVariable function is to add a variable to my contig list, like running the following:
```
contig.list_pt12_hto <- createHTOContigList(pt12, obj, group.by = c("HTO_maxID"), multi.run = "patient_experiment")

contig.list_pt12_hto <- addVariable(contig.list_pt12_hto, variable.name = "Time point", variables = c('baseline', '3 months', '6 months', '9 months', "control"))
```
This allows different contig lists (in this case the different HTO contig lists) to be annotated by the time point. However, I want to add a variable about the celltype, which is barcode level metadata. 

How could I add what celltype each barcode is from so I could visualize the data that way rather than in aggregate?  I imagine that looking by celltype is a common approach for repertoire data, so I think I am missing something.

**Comment by ncborcherding** [comment]
Hey Jasmine, 

I think what might be happening is that you have a contig list:

```
combined_bcr_pt12 <- combineBCR(contig.list_pt12_hto_celltype, samples = c("hto 1.DN2", "hto 2.DN2", "hto 3.DN2", "hto 4.DN2"))
```

This actually shouldn't work with a list of 40 contigs annotations. For ```combineBCR()``` there is a requirement for samples for each list element as there is an internal edit distance calculation. I think if you are subseting for other names using: ```combined_bcr_pt1[names(combined_bcr_pt1) %in% keep_names]``` you might be returning a list with 0 elements, which would cause an error.

You can check this with: ```names(combined_bcr_pt12)```

In terms of the ```addVariables``` - I'd do it after you have ```combineBCR()``` and make sure the length of the **variables** match the number of list elements. 

Happy to keep this thread open and continue to troubleshoot.

Nick

**Comment by jjs847** [comment]
Hi Nick, 

Thank you for the response. 

If I want to run scRepertoire on a specific cell subset -- naive B cells for example -- how could I add this **cell level variable** in order to visualize the repertoire in naive B cells only? 

I understand AddVariable lets you annotate each item in the contig list like
```contig.list <- createHTOContigList(exp1, obj, group.by = c("HTO_maxID"), multi.run = "patient_experiment")

contig.list <- addVariable(contig.list, variable.name = "condition", variables = c('disease, 'control', 'disease', 'control, "disease"))```
with the variables being a list with as many elements as the "contig.list". This allows different contig lists to be annotated by a variable. However, I want to add a variable about the celltype, which is barcode level metadata.

Best, 

Jasmine

**Comment by ncborcherding** [comment]
Hey Jasmine, 

I think probably the easiest way to do that is at the single-cell object level. A meta data variable assigning B subsets with or without the individual sample level data can then be used as a **group.by** variable for any function.

Nick

**Comment by ncborcherding** [comment]
I am going to close this issue due to inactivity, but please continue if there is anything additional.

**Comment by jjs847** [comment]
Hi, 

So sorry for the late response, I still am struggling with this issue. This is what I understood from your suggestion:
```
library(scRepertoire)

# Load the data
obj <- readRDS('/gpfs/data/ruggleslab/home/jjs847/RA_final/objects/bcell_obj_annotated.rds')
pt1 = read.csv('/gpfs/data/ruggleslab/genomics/silverman/eccite-seq/2023-01-11/cellranger/multi-Patient1/outs/per_sample_outs/multi-Patient1/vdj_b/filtered_contig_annotations.csv')

# edit rownames of cellranger outs since seurat object is integrated and has suffixes appended on to cell names
pt1$barcode <- paste0(pt1$barcode, "_1_1")

contigs <- contig.list <- createHTOContigList(pt1,
                                              obj, 
                                   group.by = "HTO_maxID")

combined.BCR <- combineBCR(contigs, 
                           samples = "pt1", 
                           threshold = 0.85)

clonalQuant(combined.BCR, 
            cloneCall="strict", 
            chain = "both", 
            scale = TRUE,
            group.by = obj$celltype)
```
I would like to look at the repertoire by annotated celltype, which is a metadata slot in my seurat object. This did not work. Do you mind providing an example on how I could accomplish this?

Best, 

Jasmine

**Comment by ncborcherding** [comment]
Hey Jasmine, 

No problem - you need to either add the clonal (```combined.BCR```) to your single-cell object or add the variables to ```combined.BCR``` itself. Here is how I would do it: 

```
obj <- combineExpression(combined.BCR, obj)
clonalQuant(obj, 
            cloneCall="strict", 
            chain = "both", 
            scale = TRUE,
            group.by = "celltype")
```
